pallets/pallet-bonded-coins/src/tests/curves/arithmetic.rs

```diff
@@ -0,0 +1,132 @@
+
+#[test]
+fn test_convert_to_fixed_basic() {
+	let x = 1000u128;
+	let denomination = 2u8; // 10^2 = 100
+
+	let result = convert_to_fixed::<Test>(x, denomination).unwrap();
+	// Test runtime uses I75F53 for CurveParameterTypeOf, which is what we'll cover
+	// in testing.
+	let expected = Float::from_num(10); // 1000 / 100 = 10
+
+	assert_eq!(result, expected);
+}
+
+#[test]
+fn test_convert_to_fixed_with_remainder() {
+	let x = 1050u128;
+	let denomination = 2u8; // 10^2 = 100
+
+	let result = convert_to_fixed::<Test>(x, denomination).unwrap();
+	let expected = Float::from_num(10.5); // 1050 / 100 = 10.5
+
+	assert_eq!(result, expected);
+}
+
+#[test]
+fn test_convert_to_fixed_smaller_than_denomination() {
+	let x = 1050u128;
+	let denomination = 6u8; // 10^6 = 1000000
+
+	let result = convert_to_fixed::<Test>(x, denomination).unwrap();
+	let expected = Float::from_num(0.00105); // 1050 / 1000000 = 0.00105
+
+	assert_eq!(result, expected);
+}
+
+#[test]
+fn test_convert_to_fixed_large_value() {
+	let x = 1_000_000_000_000_000u128;
+	let denomination = 12u8; // 10^12 = 1_000_000_000_000
+
+	let result = convert_to_fixed::<Test>(x, denomination).unwrap();
+	let expected = Float::from_num(1000); // 1_000_000_000_000_000 / 1_000_000_000_000 = 1000
+
+	assert_eq!(result, expected);
+}
+
+#[test]
+fn test_convert_to_fixed_small_denomination() {
+	let x = 12345u128;
+	let denomination = 1u8; // 10^1 = 10
+
+	let result = convert_to_fixed::<Test>(x, denomination).unwrap();
+	let expected = Float::from_num(1234.5); // 12345 / 10 = 1234.5
+
+	assert_eq!(result, expected);
+}
+
+#[test]
+fn test_convert_to_fixed_overflow() {
+	let x = u128::MAX;
+	let denomination = 0u8; // 10^0 = 1, no scaling
+
+	let result = convert_to_fixed::<Test>(x, denomination);
+	assert!(result.is_err());
+	assert_eq!(result.unwrap_err(), ArithmeticError::Overflow);
+}
+
+#[test]
+fn test_convert_to_fixed_denomination_overflow() {
+	let x = 1000u128;
+	let denomination = 128u8; // 10^128 overflows and results in division by zero
+
+	let result = convert_to_fixed::<Test>(x, denomination);
+	assert!(result.is_err());
+	assert_eq!(result.unwrap_err(), ArithmeticError::Overflow);
+}
+
+#[test]
+fn test_convert_to_fixed_overflow_avoided() {
+	let x = u128::MAX; // around 3.4e+38
+	let denomination = 17u8; // I75F53 should handle around 1.8e+22, 38 - 23 -> 17
+
+	let result = convert_to_fixed::<Test>(x, denomination);
+	assert_ok!(result);
+}
+
+#[test]
+fn test_convert_to_fixed_handles_large_denomination() {
+	let x = u128::MAX; // around 3.4e+38
+	let denomination = 22u8; // I75F53 should handle around 1.8e+22
+
+	let result = convert_to_fixed::<Test>(x, denomination);
+	assert_ok!(result);
+}
+
+#[test]
+fn test_convert_to_fixed_too_large_denomination() {
+	let x = u128::MAX; // around 3.4e+38
+	let denomination = 30u8; // I75F53 should handle around 1.8e+22, this can't be represented
+
+	let result = convert_to_fixed::<Test>(x, denomination);
+	assert_ok!(result);
+}
```

Comment:
this test fails, as a reminder that the proposed solution is still imperfect

Reply:
Due to 694334c178b2082d19b0f1e27a703585b9f865ae this test now passes; but not for u128::MAX, which results in an overflowing remainder.

Reply:
Since https://github.com/KILTprotocol/kilt-node/commit/08b71c4c8966febb0c17df070b7b3e941f77115b large denominations do not result in an overflow any more. Note that of course truncation can still happen with large denominations if the fixed type is not able to offer enough digits behind the comma.